From 9821ed36b4e5e942044b9e18c84b2efe3656c3e9 Mon Sep 17 00:00:00 2001
From: Robert Mader <robert.mader@collabora.com>
Date: Wed, 28 Jan 2026 13:49:26 +0100
Subject: [PATCH 4/6] aperture: Set video/image caps to I420 and NV12

The used format can influence the resulting video profile. Notably RGB
formats from the libcamera SW/GPU-ISP would, when used with x264enc, be
converted to Y444, i.e. a non-subsampled format, resulting in a H264
High (4/4/4) profile being used - which is undesirable for various
reasons.

Ensure we use I420, which is the default format for SDR content, and
only allow NV12 as alternative for HW-encoding.
---
 aperture/src/viewfinder.rs | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/aperture/src/viewfinder.rs b/aperture/src/viewfinder.rs
index f1431c9..265cc4f 100644
--- a/aperture/src/viewfinder.rs
+++ b/aperture/src/viewfinder.rs
@@ -315,6 +315,13 @@ mod imp {
                 .expect("Missing GStreamer Base Plug-ins");
             camerabin.set_property("video-filter", &videoconvert_video);
 
+            let caps_video = gst_video::video_make_raw_caps(&[
+                gst_video::VideoFormat::I420,
+                gst_video::VideoFormat::Nv12,
+            ])
+            .build();
+            camerabin.set_property("video-capture-caps", caps_video);
+
             let videoconvert_image = gst::ElementFactory::make("videoconvert")
                 .build()
                 .expect("Missing GStreamer Base Plug-ins");
-- 
2.52.0

